Document ppc inline asm support

pmur opened the pull request:
This PR is a sub-part of rust-lang/rust#147996.

r? @Amanieu

pmur committed:
Document ppc inline asm support

Mark status and sticky bits of vscr, fpscr and spefscr as being
preserved when using the `preserves_flags` option. These are not
enforced by codegen today, but might be in future LLVM releases.

Amanieu commented on src/inline-assembly.md:
```diff
@@ -1346,6 +1379,13 @@ r[asm.rules.preserved-registers]
+    - The `mq` register (this register only exists on POWER1 and POWER2).
+  - PowerPC SPE
+    - The sticky and status bits of the `spefscr` (any field other than FINXE, FINVE, FDBZE, FUNFE, FOVFE, or FRMC)
+    - The Accumulator register `acc` (LLVM names this `spe_acc`).
```
This should probably be defined as a clobber-only register. This should be fine to stabilize along with the other registers here.

folkertdev commented on src/inline-assembly.md:
```diff
@@ -691,6 +663,17 @@ The availability of supported types for a particular register class may depend o
+| PowerPC/PowerPC64 | `reg_nonzero` | None | `i8`, `i16`, `i32`, `i64` (PowerPC64 only) |
+| PowerPC/PowerPC64 | `freg` | None | `f32`, `f64` |
+| PowerPC/PowerPC64 | `vreg` | `altivec` | `i8x16`, `i16x8`, `i32x4`, `f32x4` |
+| PowerPC/PowerPC64 | `vreg` | `vsx` | `f32`, `f64`, `i64x2`, `f64x2` |
```
The source has the following `FIXME` linking to https://github.com/rust-lang/rust/pull/131551#discussion_r1862535963, maybe that can be resolved?

https://github.com/rust-lang/rust/blob/a3f2d5abe45a9acfaccbf09266b33e1fd7ab193e/compiler/rustc_target/src/asm/powerpc.rs#L58-L62

```rust
            // FIXME: vsx also supports integers?: https://github.com/rust-lang/rust/pull/131551#discussion_r1862535963
            Self::vreg => types! {
                altivec: VecI8(16), VecI16(8), VecI32(4), VecF32(4);
                vsx: F32, F64, VecI64(2), VecF64(2);
            },
```

cc @taiki-e

pmur replied:
The integer support for vsx is quite limited. It's only 128b integers (power8) with very restricted functionality prior to power10 (notably, no divide support). Maybe that could be supported via `VecI128(1)`? I suspect the llvm issues remain. Is that a blocker for stabilization?


F128 is another type which should be allowed, but I think that support remains unstable?

pmur replied:
> Do you think that is an oversight in LLVM?

I think so. Gcc allows it, and I believe ELFv2 calling conventions require they be passed via vector registers. LLVM also crashed when I tried wrapping it into a vector.

As for I128. It behaves strangely using LLVM. If you don't wrap it in a vector, it gets treated like a register pair. If you wrap it in  a vector and fail to set the minimum cpu to pwr8 or newer (on ppc64), it quietly generates bad code to move between gpr and vr.

folkertdev replied:
maybe our target maintainers can confirm the expected behavior here?

cc @daltenty @gilamn5tr @amy-kwan

I'd be happy to submit a patch to LLVM if we can agree on how this should work.

ehuss committed:
Adjust punctuation

traviscross committed:
Clarify verbiage in `asm.rules.stack-above-sp`

The "in specific cases" text seemed a bit unclear to me.  Let's
collapse this item with the subitem to remove this text.

traviscross committed:
Add missing period
